lib/routes/password.js

```diff
@@ -29,13 +29,239 @@ module.exports = function (
   config
   ) {
 
+  const MIN_LAST_AUTH_PASSWORD_CHANGE = config.changePassword && config.changePassword.minLastAuth || 300000 // Default to 5 mins
+
+  const totpUtils = require('../../lib/routes/utils/totp')(log, config, db)
+
   function failVerifyAttempt(passwordForgotToken) {
     return (passwordForgotToken.failAttempt()) ?
       db.deletePasswordForgotToken(passwordForgotToken) :
       db.updatePasswordForgotToken(passwordForgotToken)
   }
 
   var routes = [
+    {
+      method: 'POST',
+      path: '/password/change',
+      config: {
+        auth: {
+          strategy: 'sessionToken'
+        },
+        validate: {
+          payload: {
+            authPW: isA.string().min(64).max(64).regex(HEX_STRING).required(),
+            wrapKb: isA.string().min(64).max(64).regex(HEX_STRING).required()
+          }
+        }
+      },
+      handler: function (request, reply) {
+        log.begin('Password.change', request)
+        let devicesToNotify, verifyHash, account, newKeyFetchToken, newSessionToken
+
+        const sessionToken = request.auth.credentials
+        const originatingDeviceId = sessionToken.deviceId
+        const { wrapKb, authPW }  = request.payload
+        const ip = request.app.clientAddress
+        const wantsKeys = requestHelper.wantsKeys(request)
+        const { uid } = sessionToken
+
+        function getAccount() {
+          return db.account(uid)
+            .then((result) => account = result)
+        }
+
+        function checkStaleAuthentication() {
+          // When a user attempts to change their password, they must obtain the `wrapKb` though a reauth/login.
+          // Let's ensure that this was done within a sane amount of time.
+          const now = new Date().getTime()
+          const lastAuth = sessionToken.lastAuthAt() * 1000
+          if ((now - lastAuth) > MIN_LAST_AUTH_PASSWORD_CHANGE) {
+            throw error.staleAuthentication()
+          }
+        }
+
+        function checkTotpToken() {
+          return totpUtils.hasTotpToken({uid})
+            .then((hasToken) => {
+              // User's that have TOTP enabled, must be in a TOTP verified session to change their password
+              if (hasToken && (sessionToken.tokenVerificationId || sessionToken.authenticatorAssuranceLevel <= 1)) {
+                throw error.unverifiedSession()
+              }
```

Comment:
I wonder whether the TOTP verification should *also* have been done within a certain amount of time.  @vbudhram thoughts?